### ISSUE TITLE
Key an object in the cache by multiple keys

### ISSUE THREAD

**Issue by hithakshi** [awslabs/aws-mobile-appsync-sdk-ios#110]
**Describe the bug**
We are basically using DynamoDb in our project. It has both partition as well as hash key. 
Also we are maintaining one table for all users. 
Each user will have userID (partition key) and projectID (hash Key).
Create , update queries requires both userID and project ID. where as fetch (fetch projects) query requires only userID. 
We have setup appSync for dynamo db with queries and mutation for online and offline support. 
There's no issue in online support. 
But in offline, caching supports only single cache key. 
i.e current sdk allows only primary key to specify in cacheKeyForObject. 
appSyncClient?.apolloClient?.cacheKeyForObject = { $0["id"] } 

Also, while performing query , we can see it allows ot specify single cacheKey
                try transaction?.write(object:  object, withKey: cachKey)

**Expected behavior**
If there's a way to support combination of  partition key and hash key as primary key, it solves our problem. 

Or any workaround is also welcome.
**Environment(please complete the following information):**
 - AppSync SDK Version: [2.6.22]
 - Dependency Manager: [ Cocoapods]
 - Swift Version : [ 4.2]

**Device Information (please complete the following information):**
 - Device: [iPhone]
 - iOS Version: [iOS 12]

**Comment by palpatim** [comment]
Hi @hithakshi ,

Apollo's `cacheKeyForObject` takes a block that returns a cache key for the given object. For example, if you have an object with the following structure:

```json
type UserProject {
  userID: ID!
  projectID: ID!
  dateCreated: Int!
  foo: String
}
```

Then you could resolve the cache key with:

```swift
func cacheKeyForUserProject(userProject: MyApolloUserProjectDataType) -> String {
    return userProject.userID + "|" + userProject.projectID
}

appSyncClient?.apolloClient?.cacheKeyForObject = cacheKeyForUserProject
```

(Or however you compose the primary key)

For the second example, you might have to pre-compose the cache key:

```swift
let cacheKey = cacheKeyForUserProject(object)
try transaction?.write(object: object, withKey: cacheKey)
```

Please let me know if this addresses your questions.

**Comment by hithakshi** [comment]
Thank you for the response. I have tried the above method.  
Now records are created like "userID | projectID" format in  database file (created by Appsync in NSTemporaryDirectory). However , none of the queries are executed using offline mode. . Both results and error is nil. 
  `let query = FetchProjectQuery(userId: userID, projectID: project.id)
                self.appSyncClient?.fetch(query: query, cachePolicy: CachePolicy.returnCacheDataDontFetch, resultHandler: { (sets, error) in
                    print(sets) // both sets and error are nil
                })`

If I change cachPolicy to returnCacheDataElseFetch, then results are successfully fetched from server. 

Note: I have not used optimisticUpdate as of now. Records in database file are created by appSync internally after appsync's 'perform' action.

**Comment by artekr** [comment]
Hey @palpatim , I have exactly the same issue as @hithakshi . My query has multiple parameters, and it failed to fetch the cached object using offline mode with `CachePolicy.returnCacheDataDontFetch`.
Any updates on this issue?
Thanks

**Comment by palpatim** [comment]
@artekr This is still on my queue to investigate, but I don't have any further insights just now.

**Comment by palpatim** [comment]
@hithakshi and @artekr -- Can you both confirm that you're not using the dot character (`"."`) in your key values? Other customers are reporting (#165) that cache keys don't work using dots, and I just want to make sure that isn't part of the problems you're reporting.

**Comment by hithakshi** [comment]
Hi Tim, I am using user's email id as userId (one of the primary key) in
our project which contains dot and @ operators. (UserId, projectId) are the
primary keys. The issue other user raised with dot operators seems similar
to mine.

**Comment by amarincas** [comment]
@hithakshi I experienced the same issue when using email, because it contains the `"."` character which is also being used as separator to compose the cache key. When merging records, the cache key is being split into components and there's where the problem occurs.

**Comment by palpatim** [comment]
#186 should resolve the issue with dots in cache keys

**Comment by palpatim** [comment]
This is released on 2.10.2. Please let us know if you continue to experience problems.
